### ISSUE TITLE
Performance in parallel session Run()

### ISSUE THREAD

**Issue by makeuptransfer** [microsoft/onnxruntime#12049]
![image](https://user-images.githubusercontent.com/74530258/176676962-91e59626-6035-4262-a5a4-7209bed5ee65.png)
when i run inferenceSession::RUN() in two threads, sometimes both threads will block, if i create one session per thread, it won't happend, can u plz give me some insights, thx

**Comment by hariharans29** [comment]
Yes, that is because currently each InferenceSession only has one cuda compute stream in them and so when you run on two different threads, the compute happens only on one stream. We have a backlog item to use one compute stream per-thread and so both calls can run independently.

When you use one session per-thread you don't see this because each session is using its own compute stream now.

Linking a similar issue: https://github.com/microsoft/onnxruntime/issues/11047

CC: @pranavsharma @RyanUnderhill

**Comment by makeuptransfer** [comment]
thx for your reply, but i am using the cpu ep，is it same as cuda ep？

**Comment by makeuptransfer** [comment]
the result of perf show MlasGemmFloatKernelFma3 takes about 13% of cpu，maybe its performance related to multithread？ my intraopnums is set to one and use sequential executor

**Comment by makeuptransfer** [comment]
but when i tracing the session::run, i find the ops cost few time, but session::run itself costs a lof of time

**Comment by makeuptransfer** [comment]
@hariharans29 need help LOL

**Comment by hariharans29** [comment]
> thx for your reply, but i am using the cpu ep，is it same as cuda ep？

Oh I am sorry - I had misunderstood this to be the CUDA EP. Can you maybe share some code around how the threadpool has been configured ?

**Comment by skottmckay** [comment]
Can you please provide the code you're using to create and run the sessions, and measure performance, so that the issue can be reproduced?

The session Run is stateless, so there aren't many places where two sessions could potentially block each other.
